airbyte-integrations/connectors/source-aws-cloudtrail/source_aws_cloudtrail/source.py

```diff
@@ -0,0 +1,182 @@
+        self.client = Client(aws_key_id, aws_secret_key, aws_region_name)
+        # records_limit: is an option to limit maximum amount of records read by connector
+        # use it for testing and development porpuses only
+        self.records_left = kwargs.get("records_limit", math.inf)
+
+    def next_page_token(self, response: Mapping[str, Any]) -> Optional[Mapping[str, Any]]:
+        return response.get("NextToken")
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {"MaxResults": self.limit}
+
+        if self.start_date:
+            params["StartTime"] = self.start_date
+        if next_page_token:
+            params["NextToken"] = next_page_token
+        return params
+
+    def datetime_to_timestamp(self, date: datetime) -> int:
+        return int(datetime.timestamp(date))
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        stream_state = stream_state or {}
+        pagination_complete = False
+
+        next_page_token = None
+        while not pagination_complete:
+            params = self.request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+            try:
+                response = self.send_request(**params)
+            except botocore.exceptions.ClientError as err:
+                # returns no records if either the start time occurs after the end time
+                # or the time range is outside the range of possible values.
+                if err.response["Error"]["Code"] == "InvalidTimeRangeException":
+                    return iter(())
+                else:
+                    raise err
+
+            next_page_token = self.next_page_token(response)
+            if not next_page_token:
+                pagination_complete = True
+
+            for record in self.parse_response(response):
+                yield record
+                self.records_left -= 1
+
+                if self.records_left <= 0:
+                    # limit of fetched records is reached
+                    return iter(())
+
+        yield from []
+
+
+class IncrementalAwsCloudtrailStream(AwsCloudtrailStream, ABC):
+    @property
+    def limit(self):
+        return super().limit
+
+    # API does not support read in ascending order
+    # save state only once after full read
+    state_checkpoint_interval = None
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        # Increment record time to avoid data dupication in the next syncs
+        record_time = latest_record[self.time_field] + 1
```

Comment:
it's ok to duplicate one record. Airbyte offers at-least-once delivery guarantee not exactly-once. This will almost never happen but better be safe (not add 1, potentially duplicate records, but definitely capture all of them) than miss records

Reply:
but be aware that each new incremental sync will produce at least one duplicated record